File: cpp/2022/jan/linked-list-random-node.cc
#include "math.h"
#include "common.h"

class Solution
{
private:
    ListNode *head;

public:
    Solution(ListNode *head)
    {
        this->head = head;
    }

    int getRandom()
    {
        int res = this->head->val;
        ListNode *node = this->head->next;
        int i = 2;
        while (node)
        {
            int j = rand() % i;
            if (j == 0)
            {
                res = node->val;
            }
            i++;
            node = node->next;
        }
        return res;
    }
};

File: code/python/2020/dec/linked-list-random-node.py
import random

class Solution:
    def __init__(self, head):
        self.head = head
        self.count = 0
        while head:
            self.count += 1
            head = head.next
    
    def get_random(self):
        r = random.randint(0, self.count-1)
        n = self.head
        for _ in range(r):
            n = n.next
        return n.val